ci_diff_helper/kokoro.py

```diff
@@ -0,0 +1,353 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Set of utilities for dealing with Kokoro (Google's internal Jenkins).
+
+This module provides a custom configuration type
+:class:`Kokoro` for the `Kokoro`_ CI system.
+
+.. _Kokoro: https://www.cloudbees.com/sites/default/files/2016-jenkins-world-jenkins_inside_google.pdf
+
+This module uses a selection of environment variables to detect the state of
+Kokoro configuration. See :mod:`~ci_diff_helper.environment_vars` for more
+details.
+
+:class:`Kokoro` Configuration Type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When running in Kokoro, you can automatically detect your
+current environment and get the configuration object:
+
+.. testsetup:: auto-detect
+
+  import os
+  os.environ = {
+      'KOKORO_ENV_VAR': 'some-value',
+  }
+
+.. doctest:: auto-detect
+
+  >>> import ci_diff_helper
+  >>> config = ci_diff_helper.get_config()
+  >>> config
+  <Kokoro (active=True)>
+
+To use the :class:`Kokoro` configuration type directly:
+
+.. testsetup:: kokoro-ci-push
+
+  import os
+  os.environ = {
+      'KOKORO_ENV_VAR': 'some-value',
+      'GERRIT_BRANCH': 'master',
+      'GERRIT_CHANGE_URL': 'https://go-review.googlesource.com/94839',
+  }
+  import ci_diff_helper
+
+.. doctest:: kokoro-ci-push
+
+  >>> config = ci_diff_helper.Kokoro()
+  >>> config
+  <Kokoro (active=True)>
+  >>> config.branch
+  'master'
+  >>> config.provider
+  <KokoroRepoProvider.gerrit: 'gerrit'>
+
+During a pull request build, we can determine information about
+the current PR being built:
+
+.. testsetup:: kokoro-ci-pr
+
+  import os
+  os.environ = {
+      'KOKORO_ENV_VAR': 'some-value',
+      'GITHUB_PULL_REQUEST_NUMBER': '23',
+      'GITHUB_PULL_REQUEST_URL': (
+          'https://github.com/organization/repository/pull/23'),
+  }
+  import ci_diff_helper
+  from ci_diff_helper import _github
+
+  def mock_pr_info(slug, pr_id):
+      assert slug == 'organization/repository'
+      assert pr_id == 23
+      payload = {
+          'base': {
+              'sha': '7450ebe1a2133442098faa07f3c2c08b612d75f5',
+          },
+      }
+      return payload
+
+  _github.pr_info = mock_pr_info
+
+.. doctest:: kokoro-ci-pr
+
+  >>> config = ci_diff_helper.Kokoro()
+  >>> config
+  <Kokoro (active=True)>
+  >>> config.in_pr
+  True
+  >>> config.pr
+  23
+  >>> config.base
+  '7450ebe1a2133442098faa07f3c2c08b612d75f5'
+"""
+
+import os
+import re
+
+import enum
```

Comment:
In the world we want to live in (i.e. Python 3.6+) this **is** in the standard lib :grinning: